components/brave_ads/core/internal/common/url/url_util_internal.cc

```diff
@@ -0,0 +1,195 @@
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_ads/core/internal/common/url/url_util_internal.h"
+
+#include "base/containers/contains.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+#include "net/base/url_util.h"
+#include "url/gurl.h"
+
+namespace brave_ads::internal {
+
+namespace {
+
+constexpr char kBraveScheme[] = "brave";
+constexpr char kChromeScheme[] = "chrome";
+
+constexpr char kRewardsHostName[] = "rewards";
+constexpr char kSyncHostName[] = "sync";
+constexpr char kWalletHostName[] = "wallet";
+
+constexpr char kSettingsHostName[] = "settings";
+
+constexpr char kSearchEnginesPath[] = "/searchEngines";
+constexpr char kSearchPath[] = "/search";
+constexpr char kSearchQuery[] = "search";
+
+}  // namespace
+
+GURL ReplaceBraveSchemeWithChromeScheme(const GURL& url) {
+  if (!url.is_valid()) {
+    return url;
+  }
+
+  if (!url.SchemeIs(kBraveScheme)) {
+    return url;
+  }
+
+  GURL::Replacements replacements;
+  replacements.SetSchemeStr(kChromeScheme);
+  return url.ReplaceComponents(replacements);
+}
+
+bool ShouldSupportInternalUrl(const GURL& url) {
+  if (!url.is_valid()) {
+    return false;
+  }
+
+  // The internal brave:// scheme must be replaced with chrome:// because `GURL`
+  // does not parse the brave:// scheme.
+  const GURL modified_url = ReplaceBraveSchemeWithChromeScheme(url);
+
+  if (!modified_url.SchemeIs(kChromeScheme)) {
+    // Do not support schemes other than chrome://.
+    return false;
+  }
+
+  const std::string host_name = modified_url.host();
+
+  if (host_name == kRewardsHostName || host_name == kSyncHostName ||
+      host_name == kWalletHostName) {
+    // Support chrome://rewards, chrome://sync, and chrome://wallet hosts.
+    return true;
+  }
+
+  if (host_name != kSettingsHostName) {
+    // Do not support hosts other than chrome://settings.
+    return false;
+  }
+
+  if (modified_url.path() == kSearchEnginesPath ||
+      modified_url.path() == kSearchPath) {
+    if (!modified_url.has_query()) {
+      // Support chrome://settings/searchEngines and chrome://settings/search
+      // paths without a query.
+      return true;
+    }
+
+    return HasUrlQueryNameWithNonEmptyValue(modified_url, kSearchQuery);
+  }
+
+  // We can reject everything else!
+  return false;
+}
+
+std::optional<std::string> GetDomain(const GURL& url) {
+  if (!url.is_valid() || !url.has_host() || url.HostIsIPAddress()) {
+    return std::nullopt;  // No registry.
+  }
+
+  const size_t registry_length = GetRegistryLength(
+      url, net::registry_controlled_domains::EXCLUDE_UNKNOWN_REGISTRIES,
+      net::registry_controlled_domains::EXCLUDE_PRIVATE_REGISTRIES);
+  if (registry_length == std::string::npos || registry_length == 0) {
+    return std::nullopt;  // No registry.
+  }
+
+  const std::string host = url.host();
+
+  const size_t domain_length = host.length() - (registry_length + 1);
```

Comment:
maybe we can add `CHECK(host.length() > registry_length)`

Reply:
Fixed. I did mean to add this but forget :-)